benchmarks/CMakeLists.txt

```diff
@@ -18,18 +18,7 @@ endfunction(ConfigureBench)
 ###################################################################################################
 # - include paths ---------------------------------------------------------------------------------
 
-include_directories("${THRUST_INCLUDE_DIR}"
-                    "${CMAKE_SOURCE_DIR}"
-                    "${GTEST_INCLUDE_DIR}"
-                    "${GBENCH_INCLUDE_DIR}"
-                    "${CMAKE_CURRENT_SOURCE_DIR}")
-
-###################################################################################################
-# - library paths ---------------------------------------------------------------------------------
-
-link_directories("${CMAKE_BINARY_DIR}"
-                 "${GTEST_LIBRARY_DIR}"
```

Comment:
For my own edification: How does this work if we don't add the link directories? (Same question for include directories)

One thing that bugs me in CMake is how much stuff is implicit. How do I know I'm getting the right includes/linker commands?

Reply:
IIUC, in the implementation of the `ConfigureBench` function there is a `target_link_libraries`, that adds libraries as a dependency of the benchmark binaries.

If you have resolved these dependencies using CMake, then CMake knows whether these libraries must be built (e.g. if they are "source" dependencies) or not, and also knows what include paths, linker paths, actual libraries, etc. must be added to the binary for this to work. Or at least this is how it is supposed to work (maybe there is a different "add_dependency" command that does this all instead of "just" `target_link_dependency`).

That can be customized per benchmark binary, so it is a finer grained thing to do than just adding all possible include and library directories to all binaries, whic his what happens with the `include/link_directories` commands.

Reply:
I've turned, e.g., the `gtest` from a bare actual library (ie., just meaning `-lgtest`) into a cmake library target in the code that pulls in gtest/gbench, e.g.:

```
foreach(_lib gtest gtest_main gmock gmock_main)
  add_library(${_lib} STATIC IMPORTED)
  set_target_properties(${_lib} PROPERTIES
    IMPORTED_LOCATION "${GTEST_LIBRARY_DIR}/lib${_lib}.a"
    INTERFACE_INCLUDE_DIRECTORIES "${GTEST_INCLUDE_DIR}"
    )
endforeach()
```

This automatically happens when one builds one's own library (e.g., `add_library(rmm)`), and it will happen automatically once we use CPM or something to pull in gtest etc, too, so the above is a transitional thing.

When pulling in a cmake library targets, via `target_link_libraries(TEST gtest)`, one should think of that as expressing "TEST is using gtest". So the above will not just link to `libgtest.a` (or not even if it's a header-only lib), but also add the include directories that are part of gtest's public interface, and it could also add macros to the command line, or compiler flags (say if the lib says it needs at least cxx14). The underlying idea is to move all knowledge of what it takes to use a library to the library target, rather than to the consumer. 

That's also why there's `PRIVATE`/`PUBLIC`/`INTERFACE` to distinguish, e.g., headers that are just needed while the library is being built (`PRIVATE`, a consumer won't inherit those include paths), vs headers that are only needed when using the library as consumer (`INTERFACE`), or `PUBLIC`, which means both.